storage/src/main/java/fk/prof/storage/StorageBackedInputStream.java

```diff
@@ -0,0 +1,117 @@
+package fk.prof.storage;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * InputStream implementation which is backed by a {@link AsyncStorage} to fetch content
+ * from a location. This impl has a notion of file parts i.e. the complete file is broken
+ * into multiple parts. When it finishes reading from part 'k' it requests for the next part 'k+1'.
+ * Any exception thrown by the {@link AsyncStorage} is thrown back to the user as IOException.
+ *
+ * @see StorageBackedOutputStream
+ * @author gaurav.ashok
+ */
+public class StorageBackedInputStream extends InputStream {
```

Comment:
Assuming this class doesn't need to be thread-safe. Confirm if this is a good assumption.